os/arch/arm/src/amebad/amebad_timer_lowerhalf.c

```diff
@@ -0,0 +1,485 @@
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	DEBUGASSERT(priv);
+
+	if (priv->started) {
+		return -EPERM;
+	}
+
+	gtimer_reload(&priv->obj, timeout);
+	priv->gpt_timeout = timeout;
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: amebad_gpt_setcallback
+ *
+ * Description:
+ *   Set the user provided timeout callback.
+ *
+ * Input Parameters:
+ *   lower    - A pointer the publicly visible representation of the
+ *              "lower-half" driver state structure.
+ *   callback - The new timer expiration function pointer. If this
+ *              function pointer is NULL, then the reset-on-expiration
+ *              behavior is restored,
+ *  arg       - Argument that will be provided in the callback
+ *
+ * Returned Values:
+ *   The previous timer expiration function pointer or NULL is there was
+ *   no previous function pointer.
+ *
+ ****************************************************************************/
+static void amebad_gpt_setcallback(struct timer_lowerhalf_s *lower, tccb_t callback, void *arg)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	DEBUGASSERT(priv);
+
+	irqstate_t flags = irqsave();
+
+	/* Save the new callback */
+	priv->callback = callback;
+	priv->arg = arg;
+
+	/* If no callback is attached, the timer stop at the first interrupt */
+	if (callback != NULL && priv->started) {
+		priv->obj.handler = (void *)amebad_gpt_handler;
+		priv->obj.hid = priv;
+	} else {
+		priv->obj.handler = NULL;
+		priv->obj.hid = NULL;
+	}
+
+	irqrestore(flags);
+}
+
+/****************************************************************************
+ * Name: amebad_gpt_ioctl
+ *
+ * Description:
+ *   Any ioctl commands that are not recognized by the "upper-half" driver
+ *   are forwarded to the lower half driver through this method.
+ *
+ * Input Parameters:
+ *   lower - A pointer the publicly visible representation of the "lower-half"
+ *           driver state structure.
+ *   cmd   - The ioctl command value
+ *   arg   - The optional argument that accompanies the 'cmd'.  The
+ *           interpretation of this argument depends on the particular
+ *           command.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int amebad_gpt_ioctl(struct timer_lowerhalf_s *lower, int cmd, unsigned long arg)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	DEBUGASSERT(priv);
+
+	int ret = -EINVAL;
+
+	tmrvdbg("GPT IOCTL cmd: %d, arg: %ld\n", cmd, arg);
+
+	switch (cmd) {
+	case TCIOC_SETDIV:
+		ret = OK;
+		break;
+	case TCIOC_GETDIV:
+		ret = OK;
+		break;
+	case TCIOC_SETMODE:
+		if ((timer_mode_t) arg == MODE_FREERUN) {
+			priv->freerunmode = true;
+			priv->freeruncount = 0;
```

Comment:
you already using timer_mode_e,
then priv->mode will be better than priv->freeruncount.
Of course type is integer.

Reply:
Please refer to this link to know the reason of using 'freerunmode', 'freeruncount' and the interrupt in Free Run mode, thanks!
https://github.com/Samsung/TizenRT/pull/4328#discussion_r450589393